xds/credentials/certprovider/remote/istioca_client.go

```diff
@@ -0,0 +1,211 @@
+			CertSigner: {
+				Kind: &structpb.Value_StringValue{StringValue: c.opts.CertSigner},
+			},
+		},
+	}
+	req := &v1alpha1.IstioCertificateRequest{
+		Csr:              string(csrPEM),
+		ValidityDuration: certValidTTLInSec,
+		Metadata:         crMetaStruct,
+	}
+	ctx := metadata.NewOutgoingContext(context.Background(), metadata.Pairs("ClusterID", c.opts.ClusterID))
+	resp, err := c.client.CreateCertificate(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("create certificate: %v", err)
+	}
+
+	if len(resp.CertChain) <= 1 {
+		return nil, errors.New("invalid empty CertChain")
+	}
+
+	return resp.CertChain, nil
+}
+
+func (c *CitadelClient) getTLSDialOption() (grpc.DialOption, error) {
+	// Load the TLS root certificate from the specified file.
+	// Create a certificate pool
+	var certPool *x509.CertPool
+	var err error
+	if c.opts.TrustedRoots == nil {
+		// No explicit certificate - assume the citadel-compatible server uses a public cert
+		certPool, err = x509.SystemCertPool()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		certPool = c.opts.TrustedRoots
+	}
+	var certificate tls.Certificate
+	config := tls.Config{
+		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			if c.opts.ProvCert != "" {
+				// Load the certificate from disk
+				certificate, err = tls.LoadX509KeyPair(
+					filepath.Join(c.opts.ProvCert, "cert-chain.pem"),
+					filepath.Join(c.opts.ProvCert, "key.pem"))
+
+				if err != nil {
+					// we will return an empty cert so that when user sets the Prov cert path
+					// but not have such cert in the file path we use the token to provide verification
+					// instead of just broken the workflow
+					log.Printf("cannot load key pair, using token instead: %v", err)
+					return &certificate, nil
+				}
+				if certificate.Leaf.NotAfter.Before(time.Now()) {
+					log.Printf("cannot parse the cert chain, using token instead: %v", err)
+					return &tls.Certificate{}, nil
+				}
+			}
+			return &certificate, nil
+		},
+	}
+	config.RootCAs = certPool
+
+	// For debugging on localhost (with port forward)
+	// TODO: remove once istiod is stable and we have a way to validate JWTs locally
+	if strings.Contains(c.opts.CAEndpoint, "localhost") {
+		config.ServerName = "istiod.istio-system.svc"
+	}
+	if c.opts.CAEndpointSAN != "" {
+		config.ServerName = c.opts.CAEndpointSAN
+	}
+
+	transportCreds := credentials.NewTLS(&config)
+	return grpc.WithTransportCredentials(transportCreds), nil
+}
+
+func (c *CitadelClient) buildConnection() (*grpc.ClientConn, error) {
+	var ol []grpc.DialOption
+	var opts grpc.DialOption
+	var err error
+	if c.enableTLS {
+		opts, err = c.getTLSDialOption()
+		if err != nil {
+			return nil, err
+		}
+		ol = append(ol, opts)
+	} else {
+		opts = grpc.WithInsecure()
+		ol = append(ol, opts)
+	}
+	ol = append(ol, grpc.WithPerRPCCredentials(c.opts.TokenProvider))
+	ol = append(ol, c.opts.GRPCOptions...)
+	//security.CARetryInterceptor())
```

Comment:
we should add it back ?

https://github.com/istio/istio/blob/master/security/pkg/nodeagent/caclient/providers/citadel/client.go#L210